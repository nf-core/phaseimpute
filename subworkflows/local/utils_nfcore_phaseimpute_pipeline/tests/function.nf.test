nextflow_function {

    name "Test function phaseimpute"
    script "../main.nf"
    tag "function"

    test("Test checkFileIndex no error with empty channel") {
        function "checkFileIndex"
        tag "checkFileIndex"
        when {
            function {
                """
                input[0] = channel.fromList([
                    [[], [], []],
                    [[id: "input"], file("input.vcf"), file("input.csi")],
                    [[], [], []],
                    [[id: "input3"], file("input3.bcf"), file("input3.csi")],
                    [[], [], []],
                    [[id: "input5"], file("input5.vcf.gz"), file("input5.csi")],
                    [[], [], []],
                    [[id: "input7"], file("input7.bam"), file("input5.bai")],
                    [[], [], []],
                    [[id: "input9"], file("input9.fa"), file("input9.fai")],
                    [[], [], []]
                ])
                """
            }
        }
        then {
            assert function.success
        }
    }

    test ("Test getFileExtension") {
        function "getFileExtension"
        tag "getFileExtension"
        when {
            function {
                """
                input[0] = file("https://raw.githubusercontent.com/nf-core/test-datasets/phaseimpute/data/panel/21/1000GP.chr21.s.norel.bcf", checkIfExists: true)
                """
            }
        }
        then {
            assert function.success
            assert snapshot(function.result).match()
        }
    }

    test("Test checkFileIndex no error") {
        function "checkFileIndex"
        tag "checkFileIndex"
        when {
            function {
                """
                input[0] = channel.fromList([
                    [[id: "input"], file("input.vcf"), file("input.csi")],
                    [[id: "input2"], file("input2.vcf"), file("input2.tbi")],
                    [[id: "input3"], file("input3.bcf"), file("input3.csi")],
                    [[id: "input4"], file("input4.bcf"), file("input4.tbi")],
                    [[id: "input5"], file("input5.vcf.gz"), file("input5.csi")],
                    [[id: "input6"], file("input6.vcf.gz"), file("input6.tbi")],
                    [[id: "input7"], file("input7.bam"), file("input5.bai")],
                    [[id: "input8"], file("input8.cram"), file("input6.crai")],
                    [[id: "input9"], file("input9.fa"), file("input9.fai")],
                    [[id: "input10"], file("input10.fa"), file("input10.fai")]
                ])
                """
            }
        }
        then {
            assert function.success
        }
    }
    test("Test checkFileIndex bam bai") {
        function "checkFileIndex"
        tag "checkFileIndex"
        when {
            function {
                """
                input[0] = channel.fromList([
                    [[id: "input7"], file("input7.bam"), file("input5.csi")],
                    [[id: "input6"], file("input6.vcf.gz"), file("input6.tbi")],
                    [[id: "input8"], file("input8.bam"), file("input8.bai")]
                ])
                """
            }
        }
        then {
            assert function.failed
            assert function.stdout.contains("[id:input7]: Index file for .bam must have the extension .bai")
        }
    }
    test("Test checkFileIndex cram crai") {
        function "checkFileIndex"
        tag "checkFileIndex"
        when {
            function {
                """
                input[0] = channel.fromList([
                    [[id: "input7"], file("input7.cram"), file("input7.tbi")],
                    [[id: "input6"], file("input6.vcf.gz"), file("input6.tbi")],
                    [[id: "input8"], file("input8.bam"), file("input8.bai")]
                ])
                """
            }
        }
        then {
            assert function.failed
            assert function.stdout.contains("[id:input7]: Index file for .cram must have the extension .crai")
        }
    }
    test("Test checkFileIndex bcf csi") {
        function "checkFileIndex"
        tag "checkFileIndex"
        when {
            function {
                """
                input[0] = channel.fromList([
                    [[id: "input7"], file("input7.bcf"), file("input7.txt")],
                    [[id: "input6"], file("input6.vcf.gz"), file("input6.tbi")],
                    [[id: "input8"], file("input8.bam"), file("input8.bai")]
                ])
                """
            }
        }
        then {
            assert function.failed
            assert function.stdout.contains("[id:input7]: Index file for [.vcf, .vcf.gz, bcf] must have the extension [.tbi, .csi]")
        }
    }
    test("Test checkFileIndex vcf csi") {
        function "checkFileIndex"
        tag "checkFileIndex"
        when {
            function {
                """
                input[0] = channel.fromList([
                    [[id: "input7"], file("input7.vcf"), file("input7.bai")],
                    [[id: "input6"], file("input6.vcf.gz"), file("input6.tbi")],
                    [[id: "input8"], file("input8.bam"), file("input8.bai")]
                ])
                """
            }
        }
        then {
            assert function.failed
            assert function.stdout.contains("[id:input7]: Index file for [.vcf, .vcf.gz, bcf] must have the extension [.tbi, .csi]")
        }
    }
    test("Test checkFileIndex vcf.gz csi") {
        function "checkFileIndex"
        tag "checkFileIndex"
        when {
            function {
                """
                input[0] = channel.fromList([
                    [[id: "input7"], file("input7.vcf.gz"), file("input7.bai")],
                    [[id: "input6"], file("input6.vcf.gz"), file("input6.tbi")],
                    [[id: "input8"], file("input8.bam"), file("input8.bai")]
                ])
                """
            }
        }
        then {
            assert function.failed
            assert function.stdout.contains("[id:input7]: Index file for [.vcf, .vcf.gz, bcf] must have the extension [.tbi, .csi]")
        }
    }
    test("Test checkFileIndex fa fai") {
        function "checkFileIndex"
        tag "checkFileIndex"
        when {
            function {
                """
                input[0] = channel.fromList([
                    [[id: "input7"], file("input7.fa"), file("input7.tbi")],
                    [[id: "input6"], file("input6.vcf.gz"), file("input6.tbi")],
                    [[id: "input8"], file("input8.bam"), file("input8.bai")]
                ])
                """
            }
        }
        then {
            assert function.failed
            assert function.stdout.contains("[id:input7]: Index file for [fa, fasta] must have the extension .fai")
        }
    }
    test("Test checkFileIndex fasta fai") {
        function "checkFileIndex"
        tag "checkFileIndex"
        when {
            function {
                """
                input[0] = channel.fromList([
                    [[id: "input7"], file("input7.fasta"), file("input6.fia")],
                    [[id: "input6"], file("input6.vcf.gz"), file("input6.tbi")],
                    [[id: "input8"], file("input8.bam"), file("input8.bai")]
                ])
                """
            }
        }
        then {
            assert function.failed
            assert function.stdout.contains("[id:input7]: Index file for [fa, fasta] must have the extension .fai")
        }
    }
    test("Test Function exportCsv") {
        function "exportCsv"
        tag "exportCsv"
        when {
            params {
                outdir = "results"
            }
            function {
                """
                // define inputs of the function here. Example:
                input[0] = channel.of( [ [id:'test'], [2:"vcf",3:"index"], file("test1.vcf"), file("test2.csi") ] )
                input[1] = ["id"]
                input[2] = "id,vcf,index"
                input[3] = "impute.csv"
                input[4] = "imputation/csv"
                """
            }
        }

        then {
            assertAll(
                { assert function.success },
                { assert snapshot(function.result).match() }
            )
        }
    }
}
