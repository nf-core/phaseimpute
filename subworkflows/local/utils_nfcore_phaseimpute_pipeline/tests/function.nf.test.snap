{
    "Test Function exportCsv": {
        "content": null,
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-07-10T16:27:29.128695884"
    }
}